`"log_follow_rename": true` setting works properly for `show_file_at_commit`, `checkout_file_at_commit`

kylebebak opened the pull request:
- `"log_follow_rename": true` setting works properly for `show_file_at_commit`, `checkout_file_at_commit`
- This uses existing `filename_at_commit` function, [in the same way it's used here](https://github.com/kylebebak/GitSavvy/blob/3266859a62c12bf16da02332106e5da5ef0bfc52/core/commands/blame.py#L220-L224) in `blame_file_atcommit`
- I've tested this locally to verify both `show_file_at_commit` and `checkout_file_at_commit` work as expected, for files that have been renamed and files that haven't been renamed

kaste commented on core/commands/log.py:
```diff
@@ -250,5 +250,13 @@ def blame_file_atcommit(self):
+        commit_hash = self._commit_hash
+        file_path = self._file_path
+
+        if commit_hash and file_path and self.savvy_settings.get("log_follow_rename"):
```
At this point all three predicates should be `True` all the time if I'm reading this correctly.  

There is also https://github.com/timbrel/GitSavvy/blob/8afe2de4e8c9cb5407bd63fb649c9b604cbd5faa/core/commands/log_graph.py#L2702-L2704 which would need the same patch I think.

kylebebak replied:
Do you mean the three expressions in the `if` statement, `if commit_hash and file_path and self.savvy_settings.get("log_follow_rename"):`?

Type checker says `file_path` is `Unknown | None` in line 254 above, so I figured, and `self.filename_at_commit` takes `(str, str)`, so I figured I'd just make sure both were truthy

As for `self.savvy_settings.get("log_follow_rename")`, that's a user setting that's not guaranteed to be truthy in general, right?

---

Agreed on making same patch in `log_graph`.py` module, I'll do that

kaste replied:
> As for self.savvy_settings.get("log_follow_rename"), that's a user setting that's not guaranteed to be truthy in general, right?

Ah right, looks like `log_follow_rename` can be `False` still, but `self._commit_hash` is always truthy here and `self._file_path` too otherwise the menu item would not appear (see `def update_actions(self)`).

kaste commented on core/commands/show_file_at_commit.py:
```diff
@@ -92,7 +92,12 @@ def run(self, edit, position=None):
+            filename_at_commit = file_path
+
+        text = self.get_file_content_at_commit(filename_at_commit, commit_hash)
         render(view, text, position)
```
I don't think the changes for "show_file_at_commit" do work.  Not sure I do get them though.  What confuses me is that we get the `filename_at_commit` here but don't use it on L103 and L105 too.  Is that intentional?

Also the whole `show_file_at_commit` view has `[p]revious` and `[n]ext` handlers, starting at L201 below.  Do they just work or did you miss them?

kylebebak replied:
I realize there's an error in the code I wrote. It should be checking `self.savvy_settings.get("log_follow_rename"):`, not `self.savvy_settings.get("blame_follow_rename"):`

The code works for me in both cases, but that's because I have both of these enabled in my GitSavvy settings

```json
	"log_follow_rename": true,
	"blame_follow_rename": true,
```

I just fixed this

---

> What confuses me is that we get the filename_at_commit here but don't use it on L103 and L105 too

Thanks for catching this, I just fixed this as well

> Also the whole show_file_at_commit view has [p]revious and [n]ext handlers, starting at L201 below. Do they just work or did you miss them?

I tried making this code work and couldn't

kaste reviewed (commented):
Generally very welcome change.  I think the `show_file_at_commit` view is more complicated to do right though.  Maybe you should split this PR in two; esp. as the `checkout` part looks almost trivial.

kylebebak commented:
@kaste 

I think the bug you found in the `show_file_at_commit` code I wrote above is fixed now. The `[n]` and `[p]` hotkeys to page through commits are not working

I think having `show_file_at_commit` working for renamed files, even without the n/p hotkey support, is valuable and could go in with this PR. If you think that's essential to get these hotkeys right along with enabling `show_file_at_commit` for `log_follow_rename`, I'll make another PR

kaste commented:
I actually wonder why the defaults for `log_follow_rename` and `blame_follow_rename` are ... wrong?  At least they seem wrong.  Like you I have set them to `true`.

kaste commented:
Actually `n` and `p` should work because it also informs us if the implementation is correct here or hacky.  At least we must know why it doesn't work.)  Generally, with `show_file_at_commit`, your approach is surprising in that you patch the "refresh" command/function.  If _show_file_at_commit_ were a function I would say, shouldn't we call it with the resolved `(commit, file_name)` pair.  Isn't the resolving in the "refresh" instead of the constructor function too late and surprising?  Basically we call `show_file_at_commit(commit_hash, file_path)` but then show `commit_hash, different_file_path`.  Does this make everything easier or more complex?

Commit:
Follow renames in show_file_at_commit and gs_log

- Add `file_path_at` helper on `HistoryMixin` that gates `filename_at_commit` behind the `log_follow_rename` setting, consolidating a repeated pattern across callers.
- `LogMixin.on_highlight`: resolve the file's historical name before forwarding to `gs_show_commit_info`, so the bottom panel shows the right per-file diff for older commits.
- `gs_log_action.checkout_file_at_commit` and `gs_log_graph_action.checkout_file_at_commit`: use `file_path_at` to check out the file under its name at the target commit.
- `gs_show_file_at_commit_refresh.run`: keep `git_savvy.file_path` as the canonical HEAD name; derive `name_at_commit` for content / title and pass the canonical path to `update_reference_document`.
- `previous_file_version`: pass `follow=True` and resolve names at both `current_commit` (so `--follow` finds an existing pathspec at the range tip) and `previous_commit` (so the blob read uses the right name) — fixes the empty gutter diff at rename boundaries.
- `get_next_commit` / `get_previous_commit`: pass `follow=…`; `get_previous_commit` also resolves the name at `commit_hash` so `p`/`n` navigate across renames.
- `BlameMixin.show_file_at_commit`: stop pre-resolving the historical name when launching `gs_show_file_at_commit`; pass `self.file_path` so the view's `git_savvy.file_path` setting is consistently the HEAD name.

Commit:
- Implement rename follow caching
- Only do expensive git log --follow --name-status and parse `[(commit_hash, name_at_that_commit)]` once per file

Commit:
- Only do git log --follow if we have a pathspec
- Fix navigation bug using `n` in get_previous_commit

kylebebak commented:
Hey @kaste , I finally got around to improving things in this PR. I think it's ready for another review. The [n]ext and [p]revious behavior is working in every view/workflow in which I've tested them:

- Blame previous/next
- `git: log`, in `SHOW-COMMIT: ...` views
- `git: log current file`, in `FILE: ...` views (this now works across file renames)

I wrote part of the code with claude: here's an LLM-generated summary of the work

## Follow renames in `git: log current file` and related views

Adds a `log_follow_rename` setting (default `false`, opt-in) that lets log/diff/show-file-at-commit views trace a file across past renames instead of stopping at the rename boundary.

### What this means for users

When `log_follow_rename: true` in GitSavvy settings:

- `git: log current file`: the quick panel now lists commits from before the file was renamed, and the bottom commit-info panel shows the file's diff at each highlighted commit (previously empty for pre-rename commits)
- `SHOW-FILE-AT-COMMIT` view: `n` / `p` navigate across the rename boundary. The gutter diff (reference document) renders correctly at pre-rename commits instead of being empty
- "Checkout file at commit": checks out the file under its name at the chosen commit, not the current name
- Graph view: same fix flows through the "Checkout file at commit" / "Show file at commit" actions there too

Default behavior is unchanged when the setting is off

### Implementation

- New helper `HistoryMixin.file_path_at(file_path, commit_hash)` gates `filename_at_commit` behind `log_follow_rename` in one place, so the setting check isn't sprinkled across callers
- Canonical-name convention: `git_savvy.file_path` always holds the HEAD-side name; historical names are computed on demand. Fixed `BlameMixin.show_file_at_commit` which previously violated this by pre-resolving to the historical name when launching the `SHOW-FILE-AT-COMMIT` view
- Rename-aware navigation: `gs_show_file_at_commit_refresh.run`, `previous_file_version`, `get_previous_commit`, `get_next_commit`, and `LogMixin.on_highlight` resolve the right name at each commit involved (the current commit for `--follow` to find a valid pathspec, and the previous commit for reading the older blob)
- Performance: a naive implementation called `git log --follow {commit}.. -- file_path` per navigation event, which made every `n` / `p` press and every quick-panel highlight noticeably slow even on unrenamed files. Replaced with a single cached `file_rename_history(file_path)` walk per file (one `git log --follow --name-status -- file_path`), reducing subsequent `filename_at_commit` calls to in-memory dict lookups. First navigation per file pays one git call; everything after is free until LRU eviction
- Robustness fallbacks:
  - Merge commits with default `--diff-merges=off` emit no `--name-status` line, so they're absent from the parsed history. When a commit isn't found in the cache, `filename_at_commit` falls back to the original targeted query (also cached). This also covers commits reachable only via `next_commits`'s `branch_hint` walk that aren't in HEAD's history.
  - `--follow` requires a pathspec, so all sites that derive `follow` from a setting now also gate on `bool(file_path and ...)`. This protects the `SHOW-COMMIT` view's `n` / `p` navigation (which has no `file_path`) from a `fatal: --follow requires exactly one pathspec` regression, and applies the same defensive pattern to `gs_blame_current_file.log`.

### Caveats

- Cache is keyed on `file_path` and lives in the global LRU (size 512). If you rename a file, you need to commit this change before you can walk the file's history. Same staleness model as the existing `read_commit` / `get_file_content_at_commit` caches
- `file_rename_history` walks from HEAD, so commits reachable only via side branches fall through to the slower fallback path (correct, just one git call per such commit)

kaste commented:
Oh hi!

A lot of time passed since then and I need a primer on what we're trying to solve here.  Esp. I don't remember why we have the setting `blame_follow_rename` (and if we need it why it is `False` by default.  Same for `log_follow_rename`.  I have both set to `True`.

Then I don't know and remember what actually didn't work in the first place.  

It is okay to let AI write the code, do you still have the conversation protocol and prompts.  These are more indicative to the work as the result (sometimes).   E.g. the first commit message is just not readable.  

I need to think about if we can cache this here; i.e. are these caches safe.

kylebebak commented:
Hey @kaste , hello again =)

We're trying to solve the problem of looking at log/history of a given a file, whether through `blame` or `log` commands, and having continuity across commits that rename a file

Without these changes, this simply doesn't work. Try doing `git: log current file`, scrolling down through commits in the quick panel to a commit where the file had a different name. The output panel is empty (no diff with changes for this file in this commit). Then try doing `Show file at commit`. The output panel shows something like this

```
$ git show 0da010251:file.py

fatal: path 'file.py' exists on disk, but not in '0da010251'
```

Hitting `p` from a commit with the previous file name is a NOOP, while hitting `n` immediately jumps to the oldest commit in which the file has its "current" name

---

Just copying from an earlier comment: with the changes in this PR, the `[n]ext` and `[p]revious` behavior is working in every view/workflow in which I've tested them:

- Blame previous/next
- `git: log`, in `SHOW-COMMIT: ...` views
- `git: log current file`, in `FILE: ...` views (this now works across file renames)

And, e.g. when doing `git: log current file`, we can see the individual file diff for every commit in the quick panel, across all the names the file has ever had

---

As far as the caching goes, it's necessary to ensure decent perf. I've done a fair bit of testing and I haven't gotten any stale cache reads. I can, e.g., rename a file that's already been committed, commit the rename, rename it again, commit the second rename, and still navigate over all the commits that touched that file, in all its previous names. So yes, I think it's safe

I've attached the full prompt logs using `claude-history`, but I don't think they're very useful (too verbose). 

[conversation-2026-04-29-001009.md](https://github.com/user-attachments/files/27191207/conversation-2026-04-29-001009.md)

kaste commented:
Thanks for the claude history; I think it is readable and gives a lot of information about how that came to be.  

Also thanks for the primer of course.  

I maybe annotate some interesting claude mistakes.  Based on your findings, I implemented the same thing from a different angle.  The protocol for #2059 is here: https://pi.dev/session/#45ec33a8e0b95abd7da563beb3639203.

kaste commented on core/commands/show_file_at_commit.py:
```diff
@@ -171,13 +178,14 @@ def run(self, edit: sublime.Edit, position: Position = None, sync: bool = True)
         settings = view.settings()
         file_path = settings.get("git_savvy.file_path")
         commit_hash = settings.get("git_savvy.show_file_at_commit_view.commit")
+        name_at_commit = self.file_path_at(file_path, commit_hash)
```
This is likely the culprit for the slowness; claude calls this on the main/UI thread.  It should have put this into the `program()` which runs on a worker thread.

kylebebak replied:
`file_path_at` runs `git log --follow ...` in a subprocess under the hood. I'm pretty sure the `git ...` subprocess is the slow part of the `file_path_at`. Whether this subprocess is launched from the main thread or a worker thread won't make any difference to latency. The caching just ensures this doesn't have to be rerun every time e.g. the user scrolls to a new commit in the quick panel. The latency in this branch is already quite good for my taste, even for large repos (I've tried it on repos with ~20k commits)

kaste commented on core/git_mixins/history.py:
```diff
@@ -194,22 +194,69 @@ def get_short_hash(self, commit_hash):
+        # be the HEAD-side name, since `--follow` matches the pathspec at
+        # the tip of the range. Cached: lets callers do many cheap lookups
+        # against a single subprocess walk.
+        output = self.git(
+            "log",
+            "--follow",
+            "--name-status",
+            "--format=COMMIT %H",
+            "--", file_path
+        )
```
This is unbounded and walks the complete history which is really not ideal.

kylebebak replied:
For a given file, if you want to show its complete history, e.g. show all the commits that touched the file in the quick panel, you're going to have to walk the whole history at least once. There's no way around that. The caching ensures that for a call to e.g. `git: log current file`, we only walk this history **once**, so I actually think it's pretty close to optimal, but maybe I'm misunderstanding something

kaste commented on core/git_mixins/history.py:
```diff
@@ -194,22 +194,69 @@ def get_short_hash(self, commit_hash):
     def resolve_commitish(self, ref: str) -> str:
         return self.git("rev-parse", "--short", ref).strip()
 
+    @cached(not_if={"file_path": lambda p: not p})
```
`p: not p` is likely bogus/slop.  The functions before it should guarantee that we have a filename.

kylebebak replied:
Ok, let's remove this check if you don't think it's necessary. Or leave it there -- it doesn't cost anything

kaste commented on core/git_mixins/history.py:
```diff
@@ -194,22 +194,69 @@ def get_short_hash(self, commit_hash):
+        return history
+
     def filename_at_commit(self, filename, commit_hash):
+        # type: (str, str) -> str
+        for h, name in self.file_rename_history(filename):
+            if h.startswith(commit_hash):
+                return name
```
This is just strange because as far as I understand this, the cache is filled *once* per *filename* and then as soon as HEAD moves the chances are getting higher and higher to get a cache miss.  

You also need to guarantee that filename is a full path filename or you get clashes when working with different projects at the same time.

kylebebak replied:
When you call something like `git: log current file`, the caching that prevents repeated full walks is only done once per call. Using this command as an example, it calls `filename_at_commit`, which calls `file_rename_history`, **regardless of caching**. This calls `git log --follow ...` and walks the complete history. When scrolling through these commits, for any commit that we haven't seen yet we launch a subprocess with e.g. `['/usr/bin/git', 'show', '--no-color', '--format=fuller', '--stat', '--patch', '26beee75', '--', 'core/git_mixins/history.py']`, but we're not walking the full graph (we're looking up the file at a single commit)

As for the stale cache reads problem, thank you for calling out this possibility. I tested it as follows: I created two repos with an `index.html` file relative to the root of each, made a few commits to each repo updating `index.html`, then did `git: log current file` on both. I was unable to get any stale cache reads. I don't think they are possible across projects, even for files that have the same paths relative to the root of different repos

kaste commented on core/git_mixins/history.py:
```diff
@@ -194,22 +194,69 @@ def get_short_hash(self, commit_hash):
+        # the tip of the range. Cached: lets callers do many cheap lookups
+        # against a single subprocess walk.
+        output = self.git(
+            "log",
+            "--follow",
+            "--name-status",
+            "--format=COMMIT %H",
+            "--", file_path
+        )
+        history = []  # type: List[Tuple[str, str]]
+        current_commit = None  # type: Optional[str]
+        for line in output.splitlines():
+            if line.startswith("COMMIT "):
+                current_commit = line[7:].strip()
+            elif current_commit and line.strip():
+                cols = line.split("\t")
+                if len(cols) >= 2:
+                    history.append((current_commit, cols[-1]))
+                    current_commit = None
+        return history
+
     def filename_at_commit(self, filename, commit_hash):
+        # type: (str, str) -> str
+        for h, name in self.file_rename_history(filename):
+            if h.startswith(commit_hash):
+                return name
+        # Cache miss. Most likely a merge commit (default
+        # `--diff-merges=off` emits no name-status line and our parser
+        # drops it), or a commit reachable only via the branch_hint that
+        # `next_commits` walks but HEAD doesn't. Fall back to a targeted
+        # query so `get_file_content_at_commit` doesn't blow up.
+        return self._filename_at_commit_uncached(filename, commit_hash)
+
+    @cached(not_if={"commit_hash": is_dynamic_ref})
+    def _filename_at_commit_uncached(self, filename, commit_hash):
         # type: (str, str) -> str
         lines = self.git(
             "log",
-            "--format=",  # we don't need any commit info beside the name status
+            "--format=",
```
Surprising slop for 2026

kylebebak replied:
Surprising that it removed a comment? Should I add it back in?

kylebebak commented:
Thank you for reviewing @kaste . [I saw that you opened another PR with the same functionality](https://github.com/timbrel/GitSavvy/pull/2059). The implementation in this PR has worked well for the last couple of weeks for a couple of repos with ~20k commits. Calling it battle-tested would be a stretch (it's not even unit-tested), but I think I've run into and fixed most of the edge cases I could hit.

I don't feel qualified to review yours. I don't even have a truly deep understanding of mine. I think it would be fine to clean this one up and merge it, e.g. remove the `blame_follow_rename` setting if you think it's redundant with `log_follow_rename`, and have `log_follow_rename` default to true. Or, just merge yours. Either way I think this functionality is useful, because renaming files is fairly common, and I've been enjoying it a lot since I added it to my fork

kaste commented:
Thanks for the PR which basically showed all parts that need a change which made my PR easier to get done.  I just merged #2059 and will do some follow up PRs to get a deeper feature set.  Then release.

kylebebak commented:
Awesome, thank you @kaste =)

kylebebak commented:
@kaste , I actually just tried installing `GitSavvy` from https://github.com/timbrel/GitSavvy instead of from my fork (https://github.com/kylebebak/GitSavvy), and bugs with the history view persist. E.g. when I go back through the history of a file that's been renamed several times, things like this happen

```
During handling of the above exception, another exception occurred:

...

Traceback (most recent call last):
  File "./python3.8/contextlib.py", line 75, in inner
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/runtime.py", line 133, in task
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/runtime.py", line 359, in task
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/commands/show_commit_info.py", line 101, in run_impl
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/commands/show_commit_info.py", line 136, in _read_commit_for_file
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/fns.py", line 47, in unique
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/commands/show_commit_info.py", line 124, in <genexpr>
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/commands/show_commit_info.py", line 132, in <lambda>
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/git_mixins/history.py", line 252, in filename_at_commit
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/git_mixins/history.py", line 75, in __getitem__
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/caches.py", line 25, in __getitem__
KeyError: ('6b331c00a', '/Users/kyle/Elementary/cloud-django/sdk/inspections/views/ingest.py')
Traceback (most recent call last):
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/git_mixins/history.py", line 249, in filename_at_commit
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/git_mixins/history.py", line 75, in __getitem__
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/caches.py", line 25, in __getitem__
KeyError: ('b2464b302', '/Users/kyle/Elementary/cloud-django/sdk/inspections/views/ingest.py')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
...
KeyError: ('b2464b302', '/Users/kyle/Elementary/cloud-django/sdk/inspections/views/ingest.py')
...
KeyError: ('16ac23f31', '/Users/kyle/Elementary/cloud-django/sdk/inspections/views/ingest.py')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
...
KeyError: ('16ac23f31', '/Users/kyle/Elementary/cloud-django/sdk/inspections/views/ingest.py')
...
KeyError: ('6c188d15a1408d77713b9a1607545850bb6c7b05', '/Users/kyle/Elementary/cloud-django/sdk/inspections/views/ingest.py')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/ui__quick_panel.py", line 47, in _on_done
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/ui_mixins/quick_panel.py", line 72, in on_action_selection
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/commands/log.py", line 280, in show_file_at_commit
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/git_mixins/history.py", line 252, in filename_at_commit
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/git_mixins/history.py", line 75, in __getitem__
  File "/Users/kyle/Library/Application Support/Sublime Text/Installed Packages/GitSavvy.sublime-package/core/caches.py", line 25, in __getitem__
KeyError: ('6c188d15a1408d77713b9a1607545850bb6c7b05', '/Users/kyle/Elementary/cloud-django/sdk/inspections/views/ingest.py')
```

I don't know what the cause is, but I figure the simplest fix would be to revert your changeset and merge this one instead (I tested this one pretty heavily, on repos with tens of thousands of commits). In any case, thank you for your help

kaste commented:
I forgot to push the latest code.  Just released 2.59.0 now.

Restart after update.  🤞
